### PR TITLE
Publish our INT32, UINT32, INT64, UINT64 ASN.1 types and Z variants

### PR DESCRIPTION
**Pull request by levitte**

### CODE REVIEW

**Comment by levitte** [comment]
Note that this includes a proposed future deprecation of LONG and ZLONG

File: include/openssl/asn1t.h [hunk 1]
```diff
@@ -901,8 +901,24 @@ DECLARE_ASN1_ITEM(ASN1_FBOOLEAN)
...
+DECLARE_ASN1_ITEM(INT32)
+DECLARE_ASN1_ITEM(ZINT32)
+DECLARE_ASN1_ITEM(UINT32)
+DECLARE_ASN1_ITEM(ZUINT32)
+DECLARE_ASN1_ITEM(INT64)
+DECLARE_ASN1_ITEM(ZINT64)
+DECLARE_ASN1_ITEM(UINT64)
+DECLARE_ASN1_ITEM(ZUINT64)
+
+# if OPENSSL_API_COMPAT < 0x10200000L
```
Comment (mattcaswell) [inline include/openssl/asn1t.h]: So this deprecates from version 1.2.0 onward. I am unclear what our policy is for deprecating in binary compatible releases. Are we not allowed to do it?

Comment (levitte) [inline include/openssl/asn1t.h]: Deprecating in 1.1.*x* takes away part of the public API, and that contrary to what we say [here](https://www.openssl.org/docs/faq.html#MISC8).

**Comment by mattcaswell** [comment]
If we are deprecating LONG/ZLONG shouldn't we also remove all uses of it in our own code? I notice there are "LONG" uses in the ec, cms and dsa code, as well as ZLONG use in the dh code. In looking for that I also fell upon asn1_item_list.h which looks like it needs an update.

**Comment by levitte** [comment]
> If we are deprecating LONG/ZLONG shouldn't we also remove all uses of it in our own code? I notice there are "LONG" uses in the ec, cms and dsa code, as well as ZLONG use in the dh code. In looking for that I also fell upon asn1_item_list.h which looks like it needs an update.

Yup, I'm working through those...

**Comment by levitte** [comment]
There.

**Comment by levitte** [comment]
To experiment with this, apply this patch:

``` perl
diff --git a/Configure b/Configure
index 66541be41d..9b425a4c47 100755
--- a/Configure
+++ b/Configure
@@ -171,6 +171,7 @@ our $BSDthreads="-pthread -D_THREAD_SAFE -D_REENTRANT";
 #
 my $maxapi = "1.1.0";           # API for "no-deprecated" builds
 my $apitable = {
+    "1.2.0" => "0x10200000L",
     "1.1.0" => "0x10100000L",
     "1.0.0" => "0x10000000L",
     "0.9.8" => "0x00908000L",
```

then configure with `--api=1.2.0`

**Comment by levitte** [comment]
Note, however, that it will fail to build until #3127 has been applied...

**Comment by richsalz** [comment]
Let’s discuss this on the list?

**Comment by levitte** [comment]
Which part?  'cause there are two parts, one is simply publishing, the other is deprecating the old.  Maybe I should split this up in two PRs to avoid confusion?

**Comment by levitte** [comment]
@richsalz, I suggest you start the discussion.  I have no idea what topic you want to talk about.

File: crypto/asn1/x_long.c [hunk 2]
```diff
@@ -12,6 +12,9 @@
 #include <openssl/asn1t.h>
 #include <openssl/bn.h>
 
+#if !(OPENSSL_API_COMPAT < 0x10200000L)
```
Comment (kaduk) [inline crypto/asn1/x_long.c]: Is the idea that OPENSSL_API_COMPAT will always be used with the '<' operator?  (Should that get written down somewhere?)

Comment (levitte) [inline crypto/asn1/x_long.c]: There places where we use >=. I did it this way so the expression is kept similar to the one in asn1t.h...

**Comment by levitte** [comment]
Ping?

**Comment by levitte** [comment]
Oh, a Travis failure...

**Comment by levitte** [comment]
Ping?

**Review by richsalz — COMMENTED** [review]
Looks good to me, but perhaps @snhenson should review?  Otherwise +1 from me.

**Comment by levitte** [comment]
> Looks good to me, but perhaps @snhenson should review? Otherwise +1 from me.

I would appreciate such a review, but will not wait endlessly.

**Review by mattcaswell — APPROVED** [review]
LGTM

**Comment by levitte** [comment]
Merged.  64f11ee88883af8995d2991307ff3fb560a96fa6 to f120fa1efe6377850d1ecd389a02e0e2241912bc

**Pull request closed without merging**